### PR TITLE
🌱Avoid MachineHealtchCheck to return early on patch errors

### LINKED ISSUE #3873: Forward-port MachineHealthCheck returns immediately on patch error
#3713 improved MachineHealthCheck to returns immediately on patch error in the v0.3 branch;
These changes should now be forward ported to v0.4

/kind feature

### PR DESCRIPTION
**Pull request by arghya88**
**What this PR does / why we need it**:
Forward ports #3713 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3873

### CODE REVIEW

**Comment by arghya88** [comment]
APIdiff

```
sigs.k8s.io/cluster-api/controllers/remote
  Incompatible changes:
  - NewTestClusterCacheTracker: changed from func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName) *ClusterCacheTracker to func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName, ...string) *ClusterCacheTracker

sigs.k8s.io/cluster-api/controllers
  Compatible changes:
  - (*MachineHealthCheckReconciler).PatchHealthyTargets: added
  - (*MachineHealthCheckReconciler).PatchUnhealthyTargets: added

sigs.k8s.io/cluster-api/api/v1alpha4
  Compatible changes:
  - MachineHealthCheckStatus.RemediationsAllowed: added
  - RemediationAllowedCondition: added
+ EXIT_VALUE=1
+ set +o xtrace
```

**Comment by arghya88** [comment]
APIDiff

```
sigs.k8s.io/cluster-api/controllers
  Compatible changes:
  - (*MachineHealthCheckReconciler).PatchHealthyTargets: added
  - (*MachineHealthCheckReconciler).PatchUnhealthyTargets: added

sigs.k8s.io/cluster-api/api/v1alpha4
  Compatible changes:
  - MachineHealthCheckStatus.RemediationsAllowed: added
  - RemediationAllowedCondition: added

sigs.k8s.io/cluster-api/controllers/remote
  Incompatible changes:
  - NewTestClusterCacheTracker: changed from func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName) *ClusterCacheTracker to func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName, ...string) *ClusterCacheTracker
+ EXIT_VALUE=1
+ set +o xtrace
```

File: api/v1alpha3/machinehealthcheck_types.go [hunk 1]
```diff
@@ -93,6 +93,11 @@ type MachineHealthCheckStatus struct {
 	// +kubebuilder:validation:Minimum=0
 	CurrentHealthy int32 `json:"currentHealthy,omitempty"`
 
+	// RemediationsAllowed is the number of further remediations allowed by this machine health check before
+	// maxUnhealthy short circuiting will be applied
+	// +kubebuilder:validation:Minimum=0
+	RemediationsAllowed int32 `json:"remediationsAllowed,omitempty"`
```
Comment (sedefsavas) [inline api/v1alpha3/machinehealthcheck_types.go:99]: This was not added by this PR. If it is deleted in the main branch, we shouldn't be adding this back.
Will check more to see how this is deleted in the main branch.

Comment (arghya88) [reply api/v1alpha3/machinehealthcheck_types.go:99]: It was never there..it will be added anyway by #3884

Comment (sedefsavas) [reply api/v1alpha3/machinehealthcheck_types.go:99]: I see, then I think it'd be better to wait it merged first.

**Comment by fabriziopandini** [comment]
@arghya88 do you think you have some time to rebase this PR? I really would like to merge this forward port asap...

**Comment by arghya88** [comment]
> @arghya88 do you think you have some time to rebase this PR? I really would like to merge this forward port asap...

@fabriziopandini This PR has some code which uses some code from #3884. That is why the tests are failing. Once that PR is merged then I can rebase. Let me know if you have any alternative.

**Comment by fabriziopandini** [comment]
@arghya88 thanks for pointing this out, let's try to get #3884 merged first

**Comment by arghya88** [comment]
@sedefsavas @fabriziopandini PTAL

**Comment by fabriziopandini** [comment]
/lgtm
over to @sedefsavas for a final pass

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**